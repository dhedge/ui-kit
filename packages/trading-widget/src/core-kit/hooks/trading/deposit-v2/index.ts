export { useDeposit } from './use-deposit'
export { useDepositAllowance } from './use-deposit-allowance'
export { useDepositQuote } from './use-deposit-quote'
export { useDepositQuoteContractRead } from './use-deposit-quote-contract-read'
export { useDepositSlippage } from './use-deposit-slippage'
export { useHandlePoolDepositData } from './use-handle-pool-deposit-data'
export { useIsVaultDepositLocked } from './use-is-vault-deposit-locked'
export { useSwapDataBasedOnSendToken } from './use-swap-data-based-on-send-token'
export { useVaultDepositParams } from './use-vault-deposit-params'
export { useVaultDepositTokenAmount } from './use-vault-deposit-token-amount'
export { useVaultDepositTokens } from './use-vault-deposit-tokens'
export { useVaultDepositTransactionArguments } from './use-vault-deposit-transaction-arguments'
export { useDepositLockTime } from './use-deposit-lock-time'
export { useDepositPriceDiff } from './use-deposit-price-diff'
export { useIsDepositWithSwapTransaction } from './use-is-deposit-with-swap-transaction'
export { useMinVaultTokensReceivedAmount } from './use-min-vault-tokens-received-amount'
export { useRequiresMinDeposit } from './use-requires-min-deposit'
