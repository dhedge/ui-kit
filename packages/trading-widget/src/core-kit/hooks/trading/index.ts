export { useTradingResultHandling } from './use-trading-result-handling'
export { useHandlePoolSwapInfo } from './use-handle-pool-swap-info'
export { useTradingSettleHandler } from './use-trading-settle-handler'
export { useTradingPriceDiff } from './use-trading-price-diff'
export { useAssetPrice } from './use-asset-price'
export { useRawAssetPrice } from './use-raw-asset-price'
export { useIsTradingEnabled } from './use-is-trading-enabled'
export { useExchangeRate } from './use-exchange-rate'
export { useHandleTrade } from './use-handle-trade'
export { useMinReceiveText } from './use-min-receive-text'
export { useMaxSlippagePlaceholder } from './use-max-slippage-placeholder'
export { useSynthetixV3OraclesUpdate } from './synthetix-v3/use-synthetix-v3-oracles-update'
export { useDepositProjectedEarnings } from './projected-earnings/use-deposit-projected-earnings'
export { useProjectedEarningsCore } from './projected-earnings/use-projected-earnings-core'
export { useIsEasySwapperTrading } from './use-is-easy-swapper-trading'
export { useAvailableWithdrawLiquidity } from './synthetix-v3/use-available-withdraw-liquidity'
