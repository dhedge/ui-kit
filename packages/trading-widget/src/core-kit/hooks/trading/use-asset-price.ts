import { DEFAULT_PRECISION } from 'core-kit/const'
import type { ChainId, PoolContractCallParams } from 'core-kit/types/web3.types'
import { formatUnits } from 'core-kit/utils'

import { useRawAssetPrice } from './use-raw-asset-price'

type AssetPriceParams = PoolContractCallParams & {
  chainId: ChainId
  watch?: boolean
  disabled?: boolean
}

export const useAssetPrice = ({
  address,
  chainId,
  watch,
  disabled,
}: AssetPriceParams): string => {
  const price = useRawAssetPrice({ address, chainId, watch, disabled })
  return formatUnits(price ?? BigInt(0), DEFAULT_PRECISION)
}
