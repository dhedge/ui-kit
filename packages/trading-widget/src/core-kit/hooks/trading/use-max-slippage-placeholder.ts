import isNumber from 'lodash.isnumber'

import {
  DEFAULT_DEPOSIT_SLIPPAGE,
  DEFAULT_WITHDRAW_SLIPPAGE_SCALE,
} from 'core-kit/const'
import {
  useTradingPanelSettings,
  useTradingPanelType,
} from 'core-kit/hooks/state'

export const useMaxSlippagePlaceholder = () => {
  const [tradingType] = useTradingPanelType()
  const [{ slippage, minSlippage }] = useTradingPanelSettings()

  if (slippage !== 'auto') {
    return slippage.toString()
  }

  if (isNumber(minSlippage)) {
    return minSlippage.toString()
  }

  return tradingType === 'deposit'
    ? DEFAULT_DEPOSIT_SLIPPAGE.toString()
    : `auto ${DEFAULT_WITHDRAW_SLIPPAGE_SCALE[0]}-${
        DEFAULT_WITHDRAW_SLIPPAGE_SCALE[
          DEFAULT_WITHDRAW_SLIPPAGE_SCALE.length - 1
        ]
      }`
}
