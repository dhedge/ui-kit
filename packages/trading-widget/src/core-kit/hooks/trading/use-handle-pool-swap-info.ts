import { useEffect } from 'react'

import { usePoolStatic } from 'core-kit/hooks/pool/multicall'
import {
  useTradingPanelDepositMethod,
  useTradingPanelEntryFee,
  useTradingPanelPoolConfig,
} from 'core-kit/hooks/state'
import { useIsPoolManagerAccount } from 'core-kit/hooks/user'
import { getPercent, isBigInt } from 'core-kit/utils'

export const useHandlePoolSwapInfo = () => {
  const { address, chainId } = useTradingPanelPoolConfig()

  const {
    data: {
      easySwapperFeeDenominator,
      easySwapperFeeNumerator,
      easySwapperAllowedPools: isPoolSwapAllowed = false,
    } = {},
  } = usePoolStatic({
    address,
    chainId,
  })
  const isPoolManagerAccount = useIsPoolManagerAccount()

  const [depositMethod, setDepositMethod] = useTradingPanelDepositMethod()
  const updateEntryFee = useTradingPanelEntryFee()[1]

  useEffect(() => {
    if (isPoolManagerAccount && isPoolSwapAllowed) {
      setDepositMethod({ address, method: 'depositWithCustomCooldown' })
    }
  }, [isPoolManagerAccount, isPoolSwapAllowed, address, setDepositMethod])

  useEffect(() => {
    if (
      isBigInt(easySwapperFeeNumerator) &&
      isBigInt(easySwapperFeeDenominator) &&
      isPoolSwapAllowed &&
      depositMethod === 'depositWithCustomCooldown'
    ) {
      updateEntryFee({
        depositWithCustomCooldown: getPercent(
          Number(easySwapperFeeNumerator),
          Number(easySwapperFeeDenominator),
        ),
      })
    }
  }, [
    updateEntryFee,
    easySwapperFeeNumerator,
    easySwapperFeeDenominator,
    depositMethod,
    isPoolSwapAllowed,
  ])
}
