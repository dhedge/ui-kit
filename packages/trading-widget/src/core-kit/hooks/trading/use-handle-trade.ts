import {
  useOnTransactionEstimationError,
  useReceiveTokenInput,
  useSendTokenInput,
  useTradingPanelModal,
  useTradingPanelPoolConfig,
  useTradingPanelTransactions,
  useTradingPanelType,
} from 'core-kit/hooks/state'
import { useIsTradingEnabled } from 'core-kit/hooks/trading'
import { useIsMultiAssetWithdraw } from 'core-kit/hooks/trading/withdraw-v2/init-step'
import { useIsInsufficientBalance } from 'core-kit/hooks/user'
import { useAccount } from 'core-kit/hooks/web3'
import { EstimationError } from 'core-kit/models'
import type { ContractActionFunc } from 'core-kit/types/web3.types'
import { useTranslationContext } from 'trading-widget/providers/translation-provider'

export const useHandleTrade = (trade: ContractActionFunc) => {
  const t = useTranslationContext()
  const { account } = useAccount()
  const poolConfig = useTradingPanelPoolConfig()
  const [sendToken] = useSendTokenInput()
  const [receiveToken] = useReceiveTokenInput()
  const [type] = useTradingPanelType()
  const updateTradingModal = useTradingPanelModal()[1]
  const updatePendingTransactions = useTradingPanelTransactions()[1]
  const onTransactionEstimationError = useOnTransactionEstimationError()
  const isMultiAssetWithdraw = useIsMultiAssetWithdraw()

  const tradingEnabled = useIsTradingEnabled()
  const insufficientBalance = useIsInsufficientBalance()

  const handleTrade = async () => {
    const isDeposit = type === 'deposit'
    const chainId = poolConfig.chainId

    updateTradingModal({
      isOpen: true,
      status: 'Wallet',
      action: isDeposit
        ? 'deposit'
        : isMultiAssetWithdraw
          ? 'multi_withdraw'
          : 'single_withdraw',
      link: '',
      sendTokens: [sendToken],
      receiveTokens: isDeposit ? [receiveToken] : null,
    })

    try {
      await trade()
    } catch (error) {
      if (error instanceof EstimationError) {
        onTransactionEstimationError?.(
          error,
          poolConfig.address,
          chainId,
          account,
        )
      }
      updateTradingModal({
        isOpen: false,
        status: 'None',
        link: '',
        sendTokens: null,
        receiveTokens: null,
      })
      updatePendingTransactions({ type: 'remove', status: 'fail' })
    }
  }

  return {
    disabled: !tradingEnabled || insufficientBalance,
    label: insufficientBalance
      ? 'Insufficient Balance'
      : type === 'deposit'
        ? t.depositAction
        : isMultiAssetWithdraw
          ? t.withdrawAction
          : t.unrollAction,
    handleTrade,
  }
}
