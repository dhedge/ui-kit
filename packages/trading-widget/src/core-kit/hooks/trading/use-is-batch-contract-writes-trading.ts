import { useTradingPanelSettings } from 'core-kit/hooks/state'
import { useIsBatchContractWritesSupported } from 'core-kit/hooks/web3'

export const useIsBatchContractWritesTrading = () => {
  const isBatchContractWritesSupported = useIsBatchContractWritesSupported()
  const [{ isBatchTransactionsEnabled }] = useTradingPanelSettings()

  return isBatchContractWritesSupported && isBatchTransactionsEnabled
}
