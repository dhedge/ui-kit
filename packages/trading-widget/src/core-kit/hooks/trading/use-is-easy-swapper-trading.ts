import { useTradingPanelPoolConfig } from 'core-kit/hooks/state'

export const useIsEasySwapperTrading = () => {
  const poolConfig = useTradingPanelPoolConfig()

  return !poolConfig.usePoolLogicDeposit
}
