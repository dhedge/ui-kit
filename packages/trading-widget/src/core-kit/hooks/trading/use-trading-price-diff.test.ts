import { optimism } from 'core-kit/const'
import * as poolHooks from 'core-kit/hooks/pool'

import * as stateHooks from 'core-kit/hooks/state'
import * as tradingHooks from 'core-kit/hooks/trading'
import { TEST_ADDRESS } from 'tests/mocks'
import { renderHook } from 'tests/test-utils'

import { useTradingPriceDiff } from './use-trading-price-diff'

vi.mock('core-kit/hooks/pool', () => ({
  usePoolTokenPrice: vi.fn(),
}))

vi.mock('core-kit/hooks/state', () => ({
  useTradingPanelPoolConfig: vi.fn(),
}))

vi.mock('./use-asset-price', () => ({
  useAssetPrice: vi.fn(),
}))

describe('useTradingPriceDiff', () => {
  it('should handle zero send input value', () => {
    const sendAssetAddress = TEST_ADDRESS
    const sendAssetValue = '0'
    const receiveAssetValue = '2'
    const sendTokenPrice = '1'
    const poolTokenPrice = '2'
    const poolConfig = { address: TEST_ADDRESS, chainId: optimism.id }

    expect(Number(sendAssetValue)).toEqual(0)
    expect(Number(receiveAssetValue)).not.toEqual(0)
    expect(Number(sendTokenPrice)).not.toEqual(0)
    expect(Number(poolTokenPrice)).not.toEqual(0)

    vi.mocked(stateHooks.useTradingPanelPoolConfig).mockReturnValue(
      poolConfig as ReturnType<typeof stateHooks.useTradingPanelPoolConfig>,
    )
    vi.mocked(tradingHooks.useAssetPrice).mockReturnValue(sendTokenPrice)
    vi.mocked(poolHooks.usePoolTokenPrice).mockReturnValue(poolTokenPrice)

    const { result } = renderHook(() =>
      useTradingPriceDiff({
        sendAssetAddress,
        sendAssetValue,
        receiveAssetValue,
      }),
    )

    expect(result.current).toEqual(0)
  })

  it('should handle zero receive input value', () => {
    const sendAssetAddress = TEST_ADDRESS
    const sendAssetValue = '1'
    const receiveAssetValue = '0'
    const sendTokenPrice = '1'
    const poolTokenPrice = '2'
    const poolConfig = { address: TEST_ADDRESS, chainId: optimism.id }

    expect(Number(sendAssetValue)).not.toEqual(0)
    expect(Number(receiveAssetValue)).toEqual(0)
    expect(Number(sendTokenPrice)).not.toEqual(0)
    expect(Number(poolTokenPrice)).not.toEqual(0)

    vi.mocked(stateHooks.useTradingPanelPoolConfig).mockReturnValue(
      poolConfig as ReturnType<typeof stateHooks.useTradingPanelPoolConfig>,
    )
    vi.mocked(tradingHooks.useAssetPrice).mockReturnValue(sendTokenPrice)
    vi.mocked(poolHooks.usePoolTokenPrice).mockReturnValue(poolTokenPrice)

    const { result } = renderHook(() =>
      useTradingPriceDiff({
        sendAssetAddress,
        sendAssetValue,
        receiveAssetValue,
      }),
    )

    expect(result.current).toEqual(0)
  })

  it('should handle zero send token price', () => {
    const sendAssetAddress = TEST_ADDRESS
    const sendAssetValue = '1'
    const receiveAssetValue = '1'
    const sendTokenPrice = '0'
    const poolTokenPrice = '2'
    const poolConfig = { address: TEST_ADDRESS, chainId: optimism.id }

    expect(Number(sendAssetValue)).not.toEqual(0)
    expect(Number(receiveAssetValue)).not.toEqual(0)
    expect(Number(sendTokenPrice)).toEqual(0)
    expect(Number(poolTokenPrice)).not.toEqual(0)

    vi.mocked(stateHooks.useTradingPanelPoolConfig).mockReturnValue(
      poolConfig as ReturnType<typeof stateHooks.useTradingPanelPoolConfig>,
    )
    vi.mocked(tradingHooks.useAssetPrice).mockReturnValue(sendTokenPrice)
    vi.mocked(poolHooks.usePoolTokenPrice).mockReturnValue(poolTokenPrice)

    const { result } = renderHook(() =>
      useTradingPriceDiff({
        sendAssetAddress,
        sendAssetValue,
        receiveAssetValue,
      }),
    )

    expect(result.current).toEqual(0)
  })

  it('should handle zero receive token price', () => {
    const sendAssetAddress = TEST_ADDRESS
    const sendAssetValue = '1'
    const receiveAssetValue = '1'
    const sendTokenPrice = '1'
    const poolTokenPrice = '0'
    const poolConfig = { address: TEST_ADDRESS, chainId: optimism.id }

    expect(Number(sendAssetValue)).not.toEqual(0)
    expect(Number(receiveAssetValue)).not.toEqual(0)
    expect(Number(sendTokenPrice)).not.toEqual(0)
    expect(Number(poolTokenPrice)).toEqual(0)

    vi.mocked(stateHooks.useTradingPanelPoolConfig).mockReturnValue(
      poolConfig as ReturnType<typeof stateHooks.useTradingPanelPoolConfig>,
    )
    vi.mocked(tradingHooks.useAssetPrice).mockReturnValue(sendTokenPrice)
    vi.mocked(poolHooks.usePoolTokenPrice).mockReturnValue(poolTokenPrice)

    const { result } = renderHook(() =>
      useTradingPriceDiff({
        sendAssetAddress,
        sendAssetValue,
        receiveAssetValue,
      }),
    )

    expect(result.current).toEqual(0)
  })

  it('should calculate diff', () => {
    const sendAssetAddress = TEST_ADDRESS
    const sendAssetValue = '100'
    const receiveAssetValue = '200'
    const sendTokenPrice = '300'
    const poolTokenPrice = '400'
    const poolConfig = { address: TEST_ADDRESS, chainId: optimism.id }

    expect(Number(sendAssetValue)).not.toEqual(0)
    expect(Number(receiveAssetValue)).not.toEqual(0)
    expect(Number(sendTokenPrice)).not.toEqual(0)
    expect(Number(poolTokenPrice)).not.toEqual(0)

    vi.mocked(stateHooks.useTradingPanelPoolConfig).mockReturnValue(
      poolConfig as ReturnType<typeof stateHooks.useTradingPanelPoolConfig>,
    )
    vi.mocked(tradingHooks.useAssetPrice).mockReturnValue(sendTokenPrice)
    vi.mocked(poolHooks.usePoolTokenPrice).mockReturnValue(poolTokenPrice)

    const { result } = renderHook(() =>
      useTradingPriceDiff({
        sendAssetAddress,
        sendAssetValue,
        receiveAssetValue,
      }),
    )

    expect(result.current).toMatchSnapshot()
  })
})
