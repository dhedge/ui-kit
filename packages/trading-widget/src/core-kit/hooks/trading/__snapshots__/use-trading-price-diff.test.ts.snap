// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useTradingPriceDiff > should calculate diff 1`] = `166.66`;
